### ISSUE TITLE
segfault when glob of loop var gets freed

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#6355]
Migrated from [rt.perl.org#21469](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=21469) (status was 'resolved')

Searchable as RT21469$

**Comment by p5pRT** [comment]
### From xmath@cpan.org

#### Created by xmath@cpan.org.



If the GP of a loop var is freed inside the loop\, perl segfaults

Example&#8203;:  perl \-e 'for $x \(1\) \{ \*x = \*y \}'

Solution&#8203;: the loop context should increment the refcount of the GP

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.8.0:

Configured by xmath at Tue Jan 28 23:54:19 CET 2003.

Summary of my perl5 (revision 5.0 version 8 subversion 0) configuration:
  Platform:
    osname=darwin, osvers=6.3, archname=darwin
    uname='darwin quoose.local. 6.3 darwin kernel version 6.3: sat dec 14 03:11:25 pst 2002; root:xnuxnu-344.23.obj~4release_ppc power macintosh powerpc '
    config_args='-de -Dprefix=/usr -Dgroupcat=nidump group . -Dhostcat=nidump hosts . -Dpasscat=nidump passwd .'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-pipe -fno-common -no-cpp-precomp -fno-strict-aliasing',
    optimize='-O3',
    cppflags='-no-cpp-precomp -pipe -fno-common -no-cpp-precomp -fno-strict-aliasing'
    ccversion='', gccversion='3.1 20020420 (prerelease)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=4321
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=8
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -flat_namespace'
    libpth=/usr/lib
    libs=-lm -lc
    perllibs=-lm -lc
    libc=/usr/lib/libc.dylib, so=dylib, useshrplib=true, libperl=libperl.dylib
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dyld.xs, dlext=bundle, d_dlsymun=undef, ccdlflags=' '
    cccdlflags=' ', lddlflags=' -flat_namespace -bundle -undefined suppress'

Locally applied patches:
    


@INC for perl v5.8.0:
    /sw/lib/perl5/darwin
    /sw/lib/perl5
    /System/Library/Perl/darwin
    /System/Library/Perl
    /Library/Perl/darwin
    /Library/Perl
    /Library/Perl
    /Network/Library/Perl/darwin
    /Network/Library/Perl
    /Network/Library/Perl
    .


Environment for perl v5.8.0:
    DYLD_LIBRARY_PATH (unset)
    HOME=/Users/xmath
    LANG=en_US
    LANGUAGE (unset)
    LC_ALL=C
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/sw/bin:/sw/sbin:/bin:/sbin:/usr/bin:/usr/sbin:/usr/X11R6/bin:/Developer/Tools
    PERL5LIB=/sw/lib/perl5
    PERL_BADLANG (unset)
    SHELL=/bin/tcsh

```
</details>

**Comment by p5pRT** [comment]
### From @schwern

I confirm\.  Still a problem with 5\.8\.6 and bleadperl\.

**Comment by p5pRT** [comment]
@schwern - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @iabyn

Now fixed with commit f83b46a0147ba6f476add85d17f61a7e7fb00f21\,

which also depends on a series of previous commits that cleaned up the
iteration code&#8203;:

d57ce4dfca04719829a488a820a1248efb7e81a7 \.\.
aafca5257dfefe4cf2a531c79cadea21a566f9c3

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'open' to 'resolved'
